lib/fluent/plugin/filter_concat.rb

```diff
@@ -110,18 +110,18 @@ def process(tag, time, record)
           return flush_buffer(stream_identity)
         end
       when :regexp
-        if firstline?(record[@key])
+        case
+        when firstline?(record[@key])
           if @buffer[stream_identity].empty?
             @buffer[stream_identity] << [tag, time, record]
+            return flush_buffer(stream_identity) if lastline?(record[@key])
           else
             return flush_buffer(stream_identity, [tag, time, record])
```

Comment:
I think you may have to refactor this function. In order to handle this case (buffer not empty, new line is both start and end) properly, you need to emit _two_ logs at once: one for the previous contents of the buffer, the other for the new line.